aep/general/0162/aep.md.j2

```diff
@@ -1,5 +1,396 @@
+
+APIs **may** store a revision history for a resource. Examples of when it is
+useful include:
+
+- When it is valuable to expose older revisions of a resource via an API. This
+  can avoid the overhead of the customers having to write their own API to store
+  and enable retrieval of revisions.
+- Other resources depend on different revisions of a resource.
+- There is a need to represent the change of a resource over time.
+
+APIs implementing resources with a revision **should** abstract resource
+revisions as nested collection of the resource. Sometimes, the revisions
+collection can be a top level collection. Exceptions include:
+
+- If resource revisions are meant to have longer lifespan than the parent
+resource. In other words, resource revisions exist after resource deletion.
+
+{% tab proto %}
+
+```proto
+message BookRevision {
+  // The path of the book revision.
+  string path = 1;
+
+  // The snapshot of the book
+  Book snapshot = 2
+    [(google.api.field_behavior) = OUTPUT_ONLY];
+
+  // The timestamp that the revision was created.
+  google.protobuf.Timestamp create_time = 3
+    [(google.api.field_behavior) = OUTPUT_ONLY];
+
+  // Other revision IDs that share the same snapshot.
+  repeated string alternate_ids = 4
+    [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+
+- The `message` **must** be annotated as a resource (AIP-123).
+- The `message` name **must** be named `{ResourceType}Revision`.
+
+{% tab oas %}
+
+```json
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "type": "object",
+    "properties": {
+        "path": {
+            "type": "string"
+        },
+        "snapshot": {
+            "$ref": "#/definitions/Book",
+            "readOnly": true,
+        },
+        "create_time": {
+            "type": "string",
+            "format": "date-time",
+            "readOnly": true,
+        },
+        "alternate_ids": {
+            "type": "array",
+            "items": {
+                "type": "string"
+            },
+            "readOnly": true,
+        }
+    },
+    "required": ["path"],
+}
+```
+
+{% endtabs %}
+
+- The resource revision **must** contain a field with a message type of the
+  parent resource, with a field name of `snapshot`.
+    - The value of `snapshot` **must** be the configuration of the parent
+      at the point in time the revision was created.
+- The resource revision **must** contain a `create_time` field (see [AIP-142][]).
+- The resource revision **may** contain a repeated field `alternate_ids`, which would
+  contain a list of resource IDs that the revision is also known by (e.g. `latest`)
+
+### Creating Revisions
+
+Depending on the resource, different APIs may have different strategies for
+creating a new revision. Specifically examples of strategies include:
+
+- Creating a revision when there is a change to the parent resource
+- Creating a revision when important system state changes
+- Creating a revision when specifically requested
```

Comment:
one things to note is that I do believe that this provision allows one to create a revision object with a create method, with a user-settable id. The value would just be the id, with the snapshot being whatever the resource was at the time the revision was created.

This does allow the resource to be managed via Declarative clients, although I'm not 100% sure what the value would be there (perhaps if constructing a resource requires that it references a revision and not the actual resource).

Reply:
there's no clause that says you *can't* do that - and actually it's a good missing example! I'll add it as one of the possible times to create a revision.

Do we need to go as far as requiring at least one revision is created? doesn't seem strictly necessary, but I imagine most cases for resource revisions would benefit from one always being created and can therefore be referenced.

Reply:
Yeah, I think it's a bit weird to have a resource with no revisions - kind of like a git repo with no commits.  I'm open to being persuaded otherwise but my current inclination is that the list of revisions should never be empty.

Reply:
If a company comes along and they want to adopt the AEPs, and they have a resource with no revisions, would we block their adoption of AEPs based on this reason? 

I'd argue no: although I see a minor point on usability, I don't see a reason why this makes a client less usable.

I'd be open to make this a should though?

Reply:
I think it would be suboptimal if users of multiple AEP-compliant APIs from different organizations had to learn this kind of difference between the two.  I agree it probably doesn't break generated clients.

Maybe a **should** is the right balance.

Reply:
SGTM, added.